pageserver/src/thread_mgr.rs

```diff
@@ -0,0 +1,297 @@
+///
+/// Any None arguments are ignored. For example, to check if there are any WalReceiver
+/// threads running:
+///
+///   is_thread_running(Some(ThreadKind::WalReceiver), None, None)
+///
+/// Or to check if there are any threads for given timeline runnning:
+///
+///   is_thread_running(None, Some(timelineid), None)
+///
+pub fn is_thread_running(
+    kind: Option<ThreadKind>,
+    tenant_id: Option<ZTenantId>,
+    timeline_id: Option<ZTimelineId>,
+) -> bool {
+    let threads = THREADS.lock().unwrap();
+    for thread in threads.values() {
+        if kind.is_none() || thread.kind == kind.unwrap() {
+            let inner_guard = thread.inner.lock().unwrap();
+            if (tenant_id.is_none() || inner_guard.tenant_id == tenant_id)
+                && (timeline_id.is_none() || inner_guard.timeline_id == timeline_id)
+            {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+/// Signal and wait for threads to shut down.
+///
+/// The arguments are used to select the threads to kill, like in `is_thread_running`.
+pub fn shutdown_threads(
+    kind: Option<ThreadKind>,
+    tenant_id: Option<ZTenantId>,
+    timeline_id: Option<ZTimelineId>,
+) {
+    let mut victim_threads = Vec::new();
+
+    let threads = THREADS.lock().unwrap();
+    for thread in threads.values() {
+        if kind.is_none() || thread.kind == kind.unwrap() {
+            let inner_guard = thread.inner.lock().unwrap();
+            if (tenant_id.is_none() || inner_guard.tenant_id == tenant_id)
+                && (timeline_id.is_none() || inner_guard.timeline_id == timeline_id)
+            {
+                thread.shutdown_requested.store(true, Ordering::Relaxed);
+                // FIXME: handle error?
+                let _ = thread.shutdown_tx.send(());
+                victim_threads.push(Arc::clone(thread));
+            }
+        }
+    }
+    drop(threads);
+
+    for thread in victim_threads {
+        info!("waiting for {} to shut down", thread.name);
+        if let Some(join_handle) = thread.inner.lock().unwrap().join_handle.take() {
+            let _ = join_handle.join();
+        } else {
+            // The thread had not even fully started yet. Or it was shut down
+            // concurrently and alrady exited
+        }
+    }
+}
+
+///
+/// Remember that the current thread is operating on the given tenant and timeline.
+///
+pub fn associate_thread(tenant_id: Option<ZTenantId>, timeline_id: Option<ZTimelineId>) {
+    CURRENT_THREAD.with(|ct| {
+        let ct = ct.borrow_mut();
+        let mut inner_guard = ct.as_ref().unwrap().inner.lock().unwrap();
+
+        inner_guard.tenant_id = tenant_id;
+        inner_guard.timeline_id = timeline_id;
+    })
+}
+
+/// A Future that can be used to check if the current thread has been requested to
+/// shut down.
+pub async fn shutdown_requested() {
+    let _ = SHUTDOWN_RX
+        .with(|rx| rx.borrow().as_ref().unwrap().clone())
```

Comment:
I'm not fond of `unwrap()` in general, so I wonder, would it make sense to either
* avoid panicking (log and error and move on?)
* panic with `context` that explains what invariant did we break so hard that we panic

Sure verbose things like `lock().unwrap()` are exception, but `Option::unwrap()` is already a nice thing to document IMO.

Reply:
Changed it to use `Option::expect`, so you get a panic with a better message.